### PR TITLE
initial logging for route protection

### PR DESCRIPTION
**Pull request by venusbb**
Continue from PR #1365

This PR retrieve from environment variable and test it again the x.44 header; and log the results.

**Related PR:** 
https://github.com/alphagov/notifications-api/pull/1365

**pivotal story:**
https://www.pivotaltracker.com/story/show/146688097

### CODE REVIEW

File: tests/app/authentication/test_authentication.py [hunk 1]
```diff
@@ -372,3 +372,66 @@ def test_allow_valid_ips_bits(restrict_ip_sms_app):
...
+        headers=[
+            ('X-Custom-forwarder', 'some key 1'),
+            ('X-Forwarded-For', '200.200.200.222, 222.222.222.222, 127.0.0.1'),
+        ]
+    )
+    assert response.status_code == 200
+
+
+@pytest.fixture
+def route_secret_app_2():
```
Comment (leohemsted) [inline tests/app/authentication/test_authentication.py]: dont think you need this second blueprint

File: tests/app/authentication/test_authentication.py [hunk 2]
```diff
@@ -372,3 +372,66 @@ def test_allow_valid_ips_bits(restrict_ip_sms_app):
...
+def test_can_use_secret_route_key_2(route_secret_app_2):
+
+    response = route_secret_app_2.get(
+        path='/',
+        headers=[
+            ('X-Custom-forwarder', 'key_2'),
+            ('X-Forwarded-For', '200.200.200.222, 222.222.222.222, 127.0.0.1'),
+        ]
+    )
+    assert response.status_code == 200
```
Comment (leohemsted) [inline tests/app/authentication/test_authentication.py]: can you try `mock_log = mocker.patch('app.authentication.auth.current_app.logger.info')`

and then `mock_log.assert_called_once_with({....})`

see if you can get that to work? that'll help make these tests more useful

File: tests/app/authentication/test_authentication.py [hunk 3]
```diff
@@ -372,3 +372,121 @@ def test_allow_valid_ips_bits(restrict_ip_sms_app):
     )
 
     assert response.status_code == 200
+
+
+@pytest.fixture
+def route_secret_app_with_key_1_only():
```
Comment (richardc0) [inline tests/app/authentication/test_authentication.py]: A test with neither key1 or 2?

File: tests/app/test_cloudfoundry_config.py [hunk 4]
```diff
@@ -16,7 +16,9 @@ def notify_config():
             'admin_client_secret': 'admin client secret',
             'secret_key': 'secret key',
             'dangerous_salt': 'dangerous salt',
-            'allow_ip_inbound_sms': ['111.111.111.111', '100.100.100.100']
+            'allow_ip_inbound_sms': ['111.111.111.111', '100.100.100.100'],
+            'route_secret_key_1': "key_1",
```
Comment (richardc0) [inline tests/app/test_cloudfoundry_config.py]: Should you test four configs?

- key 1 only
- key 2 only
- key 1 and key 2
- no keys

File: app/authentication/auth.py [hunk 5]
```diff
@@ -44,11 +45,41 @@ def requires_no_auth():
...
 
 
-def restrict_ip_sms():
+def check_route_secret():
     # Check route of inbound sms (Experimental)
     # Temporary custom header for route security
-    if request.headers.get("X-Custom-forwarder"):
-        current_app.logger.info("X-Custom-forwarder received")
+    if request.headers.get("X-Custom-Forwarder"):
+        route_secret_key = request.headers.get("X-Custom-Forwarder")
```
Comment (richardc0) [inline app/authentication/auth.py]: Test for secret_key is None?

File: app/authentication/auth.py [hunk 6]
```diff
@@ -44,11 +45,41 @@ def requires_no_auth():
...
+
+        # if route_secret_key is None:
```
Comment (richardc0) [inline app/authentication/auth.py]: Removed commented out code

File: tests/app/test_cloudfoundry_config.py (hunk 4 above)
Comment (venusbb) [reply tests/app/test_cloudfoundry_config.py]: Case 1-3 are addressed in test_authentication.py. Will build logic around case 4

File: app/authentication/auth.py (hunk 6 above)
Comment (venusbb) [reply app/authentication/auth.py]: We need this when we enable blocking once we check everything is OK.

File: tests/app/authentication/test_authentication.py (hunk 3 above)
Comment (venusbb) [reply tests/app/authentication/test_authentication.py]: Modified auth.py to return error. Added a new fixture and test.

File: app/authentication/auth.py [hunk 7]
```diff
@@ -47,8 +47,27 @@ def requires_no_auth():
...
+    if request.headers.get("X-Custom-Forwarder"):
+        route_secret_key = request.headers.get("X-Custom-Forwarder")
+        key_1 = current_app.config.get('ROUTE_SECRET_KEY_1')
+        key_2 = current_app.config.get('ROUTE_SECRET_KEY_2')
+        key_used = 0
+        route_allowed = False
+        if route_secret_key == key_1:
+            key_used = 1
+            route_allowed = True
+        elif route_secret_key == key_2:
```
[the remark this answers is not in the document]
Comment (venusbb) [reply app/authentication/auth.py]: else it is false, which is set before the check.

**Pull request merged**